src/buffer/out/textBuffer.cpp

```diff
@@ -376,6 +376,24 @@ bool TextBuffer::_PrepareForDoubleByteSequence(const DbcsAttribute dbcsAttribute
     return fSuccess;
 }
 
+void TextBuffer::ConsumeGrapheme(std::wstring_view& chars) noexcept
+{
+    // This function is supposed to mirror the behavior of ROW::Write, when it reads characters off of `chars`.
+    // (I know that a UTF-16 code point is not a grapheme, but that's what we're working towards.)
+    chars = til::utf16_pop(chars);
+}
+
+void TextBuffer::Write(til::CoordType row, bool wrapAtEOL, const TextAttribute& attributes, RowWriteState& state)
+{
+    auto& r = GetRowByOffset(row);
+
+    r.ReplaceText(state);
+    r.ReplaceAttributes(state.columnBegin, state.columnEnd, attributes);
+    r.SetWrapForced(wrapAtEOL && state.columnEndDirty >= r.size());
+
+    TriggerRedraw(Viewport::FromExclusive({ state.columnBegin, row, state.columnEnd, row + 1 }));

```

Comment:
This may not matter if we're doing full row redraws anyway, but shouldn't this be  using `columnEndDirty`?

Reply:
Oh that's interesting... It should also be `state.columnBeginDirty`! But that code behaves identical to the current state (before this PR). Unless I'm mistaken, I think this is a regression due to #13626, which started to properly clear leading/trailing half of wide glyphs and this `TriggerRedraw` should've been fixed in that PR.